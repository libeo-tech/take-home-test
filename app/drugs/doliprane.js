import {DrugModel} from "./models/drug";

export class Doliprane extends DrugModel {
  constructor(expiresIn, benefit) {
    super(DrugModel.drugs.DOLIPRANE, expiresIn, benefit)
  }

  updateBenefitValue(data) {
    try {
      DrugModel.validate(data);

      return DrugModel.updateBenefitValue(data);
    } catch (error) {
      throw error;
    }
  }

  updateExpiresInValue(data) {
    try {
      DrugModel.validate(data);

      return DrugModel.updateExpiresInValue(data);
    } catch (error) {
      throw error;
    }
  }
}
