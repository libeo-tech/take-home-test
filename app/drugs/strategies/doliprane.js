import {Drug} from "../drug";
import {DrugStrategy} from "./drug-strategy";

export class DolipraneStrategy extends DrugStrategy {
  constructor(name) {
    super();
    this.name = name;
    this.updateBenefitValue = this._updateBenefitValue;
    this.updateExpiresInValue = this._updateExpiresInValue;
  }

  _updateBenefitValue(data) {
    try {
      Drug.validate(data);

      if (data.benefit == Drug.minBenefitValue) return data;

      let benefitReduction = 0;
      if (data.benefit > Drug.minBenefitValue) benefitReduction++;
      if (data.expiresIn <= 0) benefitReduction++;

      data.benefit = data.benefit - benefitReduction;

      if (data.benefit < Drug.minBenefitValue) data.benefit = Drug.minBenefitValue;

      return data;
    } catch (error) {
      throw error;
    }
  }

  _updateExpiresInValue(data) {
    try {
      Drug.validate(data);

      return DrugStrategy.updateExpiresInValue(data);
    } catch (error) {
      throw error;
    }
  }
}
