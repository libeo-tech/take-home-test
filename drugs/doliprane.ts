import { Drug, DrugName } from "./drug";

export class Doliprane extends Drug {
    constructor(expiresIn: number, benefit: number) {
        super(DrugName.DOLIPRANE, expiresIn, benefit);
    }

    public updateDrug(): void {
        if (this.isExpired) {
            this.resetBenefit();
            return;
        }
       this.decreaseExpiresIn();
       this.decreaseBenefit();
    }
}