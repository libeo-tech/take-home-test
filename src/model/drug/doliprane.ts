import { DrugName } from '../../type/DrugName';
import { Drug } from './drug';

export class Doliprane extends Drug {
  constructor(expiresIn: number, benefit: number) {
    super(DrugName.DOLIPRANE, expiresIn, benefit);
  }
}
