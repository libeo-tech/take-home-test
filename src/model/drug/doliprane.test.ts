import { Doliprane } from './doliprane';

describe('Doliprane', () => {
  let doliprane: Doliprane;

  beforeEach(() => {
    doliprane = new Doliprane(30, 5);
  });

  it('should not allow for a negative benefit', () => {
    doliprane.benefit = -5;
    expect(doliprane).toMatchSnapshot();
  });

  it('should not allow for a benefit greater than 50', () => {
    doliprane.benefit = 51;
    expect(doliprane).toMatchSnapshot();
  });

  it('should return expiration status', () => {
    expect(doliprane.isExpired()).toBe(false);
    doliprane.expiresIn = 0;
    expect(doliprane.isExpired()).toBe(true);
  });

  it('should decrease expiresIn and benefit over time', () => {
    doliprane.updateValues();
    expect(doliprane).toMatchSnapshot();
  });

  it('should decrease benefit twice as much if expired', () => {
    doliprane.expiresIn = 0;
    doliprane.updateValues();
    expect(doliprane).toMatchSnapshot();
  });

  it('should be serializable to JSON', () => {
    expect(doliprane.toJSON()).toMatchSnapshot();
  });

  it('should be serializable to string', () => {
    expect(doliprane.toString()).toMatchSnapshot();
  });
});
