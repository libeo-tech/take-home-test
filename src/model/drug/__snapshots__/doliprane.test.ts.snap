// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Doliprane should be serializable to JSON 1`] = `
Object {
  "benefit": 5,
  "expiresIn": 30,
  "name": "Doliprane",
}
`;

exports[`Doliprane should be serializable to string 1`] = `"{\\"name\\":\\"Doliprane\\",\\"expiresIn\\":30,\\"benefit\\":5}"`;

exports[`Doliprane should decrease benefit twice as much if expired 1`] = `
Object {
  "benefit": 3,
  "expiresIn": -1,
  "name": "Doliprane",
}
`;

exports[`Doliprane should decrease expiresIn and benefit over time 1`] = `
Object {
  "benefit": 4,
  "expiresIn": 29,
  "name": "Doliprane",
}
`;

exports[`Doliprane should not allow for a benefit greater than 50 1`] = `
Object {
  "benefit": 50,
  "expiresIn": 30,
  "name": "Doliprane",
}
`;

exports[`Doliprane should not allow for a negative benefit 1`] = `
Object {
  "benefit": 0,
  "expiresIn": 30,
  "name": "Doliprane",
}
`;
